src/dpo/src/Optdp.cpp

```diff
@@ -0,0 +1,855 @@
+// BSD 3-Clause License
+//
+// Copyright (c) 2021, Andrew Kennings
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "dpo/Optdp.h"
+
+#include <odb/db.h>
+#include <boost/format.hpp>
+#include <cfloat>
+#include <cmath>
+#include <iostream>
+#include <limits>
+#include <map>
+
+#include "dpl/Opendp.h"
+#include "ord/OpenRoad.hh"  // closestPtInRect
+#include "utl/Logger.h"
+
+// My stuff.
+#include "architecture.h"
+#include "detailed.h"
+#include "detailed_manager.h"
+#include "legalize_shift.h"
+#include "network.h"
+#include "router.h"
+
+namespace dpo {
+
+using std::round;
+using std::string;
+
+using utl::DPO;
+
+using odb::dbBlock;
+using odb::dbBox;
+using odb::dbBPin;
+using odb::dbBTerm;
+using odb::dbITerm;
+using odb::dbMasterType;
+using odb::dbMPin;
+using odb::dbMTerm;
+using odb::dbNet;
+using odb::dbPlacementStatus;
+using odb::dbRegion;
+using odb::dbSBox;
+using odb::dbSet;
+using odb::dbSigType;
+using odb::dbSWire;
+using odb::dbTechLayer;
+using odb::dbWireType;
+using odb::Rect;
+
+static bool swapWidthHeight(dbOrientType orient);
+
+////////////////////////////////////////////////////////////////
+Optdp::Optdp() : nw_(nullptr), arch_(nullptr), rt_(nullptr) {}
+
+////////////////////////////////////////////////////////////////
+Optdp::~Optdp() {}
+
+////////////////////////////////////////////////////////////////
+void Optdp::init(ord::OpenRoad* openroad) {
```

Comment:
It is better to take the pieces of the app you actually need as args rather than the whole thing (ie db, logger, opendp).